Literal::Array: implement `#replace`, and `#insert`. add `#prepend` alias

phillipspc opened the pull request:
Continuing to chip away at #134 
*** 
My goal was to more or less work my way down the list of "Methods for Assigning" from the ruby docs
<img width="808" alt="CleanShot 2024-11-11 at 14 20 07@2x" src="https://github.com/user-attachments/assets/93e37daa-4dda-4cd2-9c06-dc910bc8417c">

From that list, this implements:
- `#prepend` (alias for `#unshift`)
- `#insert`
- `#replace`

I'm still not 100% on if I'm always using `Literal.check` correctly. And also unsure how thorough to be in tests. Feedback welcome of course!

phillipspc commented on lib/literal/array.rb:
```diff
@@ -196,6 +218,24 @@ def reject!(...)
 		self
 	end
 
+	def replace(value)
```
I'm wondering if this should work more like map where you could say replace a `Literal::Array(Integer)` with a `Literal::Array(String)` ? 🤔

joeldrapper replied:
I think we should avoid any type-mutating mappings. We should be able to consider the type of a Literal::Array frozen even if the contents are mutable (within the constraints of the type).

We can skip checking types on a `Literal::Array(@__type__)` here since it’s already the same type.

phillipspc replied:
To clarify my comment, in the issue I see following example code
```ruby
array = Literal::Array(Integer).new(1, 2, 3)
array.map(String, &:to_s) # ["1", "2", "3", "4"]
```
So that's kind of where I was getting this notion of type mutating from (when that new type is specified). But if you're saying we don't want to do that here, that's fine by me 👍

joeldrapper replied:
I like this for anything that returns a new object, but not for things that mutate the object. I think it’s useful to know if you’ve got a Literal::Array(Integer), it’s never going to turn into a Literal::Array(String), even if things might be added or removed form it.

joeldrapper commented on lib/literal/array.rb:
```diff
@@ -106,6 +106,19 @@ def clear(...)
 		self
 	end
 
+	def concat(*value)
+		unless [Array, Literal::Array].include?(value.flatten(1).class)
+			raise ArgumentError.new("#concat expects Array arguments")
+		end
+
+		Literal.check(actual: value.flatten(1), expected: @__collection_type__) do |c|
+			c.fill_receiver(receiver: self, method: "#concat")
+		end
+
+		@__value__.concat(*value)
+		self
+	end
```
I think we can improve this with a separate case for Literal::Array. Something like

```ruby
case
when Array
  # we need to type check here
when Literal::Array(@__type__)
  # we don’t actually need to type check because we’re guaranteed the other array has the same type as this one so we can just `@__value__.concat(other.__value__)`.
when Literal::Array
  # here, we can actually do some more checks to avoid type checking but let’s just type check for now. Could you leave a todo for me?
end
```

joeldrapper replied:
Wait, `concat` takes `*value` so it’s always an Array, never a Literal::Array. 🤔

phillipspc replied:
Yeah because of the splat, `value` is always an array, and I was checking what was in _that_ array with `value.flatten(1)`

joeldrapper replied:
But it’s mean to be an array of arrays so we can handle them without the type check.

phillipspc replied:
We discussed this elsewhere, but since `concat` presents some particular challenges, I've pulled it out of this PR for now 👍

joeldrapper commented on lib/literal/array.rb:
```diff
@@ -196,6 +218,24 @@ def reject!(...)
+						when Array
+							value
+						when Literal::Array
+							value.to_a
```
I don’t know if `Array#replace(value)` will mutate the given value or just use it as a reference. If it doesn’t mutate, we could call `.__value__` here instead of `.to_a`.

phillipspc replied:
I don't  _think_ value gets mutated but not 100% sure the best way to check
<img width="494" alt="CleanShot 2024-11-11 at 15 52 31@2x" src="https://github.com/user-attachments/assets/5dc01a6d-84ea-445c-8e29-24a6883a2cc5">

phillipspc commented:
@joeldrapper Based on our conversation, I've removed the `concat` attempt for now 😅 . I've also updated `replace` to hopefully be more in line with what you were looking for. Let me know what you think

joeldrapper commented on lib/literal/array.rb:
```diff
@@ -196,6 +205,27 @@ def reject!(...)
 
+	def replace(value)
+		case value
+		when Array
```
Does this need to be `::Array`? I’m not sure if it’s going to resolve to itself here since the class is called `Array`.
